lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/TxHistory.hs

```diff
@@ -85,3 +114,75 @@ instance Monoid TxHistory where
 
 type ResolveAddress = Address -> Maybe Customer
 type ResolveSlot = Slot -> Maybe DownTime
+type ResolveValueTransfer =
+    forall era. IsEra era => Tx era -> Map.Map Address ValueTransfer
+
+getEraSlotOfBlock :: IsEra era => Block era -> Slot
+getEraSlotOfBlock = At . getEraSlotNo . getEraBHeader
+
+-- | Roll forward a block into the transaction history. This function
+-- relies on the TxHistory to be sorted by time both on the time and
+-- customer views.
+rollForward
+    :: ResolveValueTransfer
+    -> ResolveAddress
+    -> ResolveSlot
+    -> EraValue Block
+    -> TxHistory
+    -> TxHistory
+rollForward resolveValueTransfer resolveAddress resolveSlot block =
+    (<> txHistory')
+  where
+    txHistory' =
+        applyEraFun
+            (blockToTxHistory resolveValueTransfer resolveAddress resolveSlot)
+            block
+
+blockToTxHistory
+    :: IsEra era
+    => ResolveValueTransfer
+    -> ResolveAddress
+    -> ResolveSlot
+    -> Block era
+    -> TxHistory
+blockToTxHistory resolveValueTransfer resolveAddress resolveSlot block =
+    fold $ do
+        tx <- getEraTransactions block
+        let slot = getEraSlotOfBlock block
+        time <- maybeToList $ resolveSlot slot
+        (address, valueTransfer) <- Map.toAscList $ resolveValueTransfer tx
+        customer <- maybeToList $ resolveAddress address
+        let byTime =
+                singletonFinger () time
+                    $ singletonMap (firstJust slot) customer
+                    $ singletonMap (firstJust address) (getTxId tx)
+                    $ Value valueTransfer
+        let byCustomer =
+                singletonMap () customer
+                    $ singletonFinger (firstJust address) time
+                    $ singletonMap (firstJust slot) (getTxId tx)
+                    $ Value valueTransfer
+        pure $ TxHistory{byCustomer, byTime}
+
+-- | Roll backward the transaction history to a given slot. This function
+-- relies on the TxHistory to be sorted by time both on the time and
+-- customer views.
+rollBackward
+    :: ResolveSlot
+    -> Slot
+    -> TxHistory
+    -> TxHistory
+rollBackward
+    resolveSlot
+    slot
+    TxHistory
+        { byCustomer
+        , byTime
+        } =
+        TxHistory
+            { byCustomer = onMap byCustomer $ fmap (`onFinger` takeToSlot)
+            , byTime = onFinger byTime takeToSlot
+            }
+        where
+            takeToSlot :: Monoid a => TimedSeq DownTime a -> TimedSeq DownTime a
+            takeToSlot x = maybe x (`before` x) (resolveSlot slot)
```

Comment:
I'm not sure whether this has an off-by-one error or not: The slot to which we roll back needs to be included in the new `TxHistory`, but the name `before` suggests that it is not included (only data "before" the slot is kept, not data "at" the slot). My personal favorite for the name of the function that does the right thing is "`dropAfter`".